src/validate.js

```diff
@@ -0,0 +1,198 @@
+
+const validator = {
+    tests: {
+        required: (value) => value !== '',
+        email: (value) => value === '' || EMAIL_REGEX.test(value),
+        minlength: (value, length) => value === '' || value.length >= parseInt(length),
+        maxlength: (value, length) => value === '' || value.length <= parseInt(length),
+        numeric: (value) => value === '' || (!isNaN(parseFloat(value)) && isFinite(value)),
+        integer: (value) => value === '' || (!isNaN(value) && !isNaN(parseInt(value, 10)) && Math.floor(value) == value), // eslint-disable-line eqeqeq
+        min: (value, min) => value === '' || parseFloat(value) >= parseFloat(min),
+        max: (value, max) => value === '' || parseFloat(value) <= parseFloat(max),
+        pattern: (value, pattern) => value === '' || (new RegExp(pattern)).test(value),
+        equals: (value, otherValue) => value === '' || value === otherValue,
+        minoptions: (value, min) => value === '' || (Array.isArray(value) && value.length >= parseFloat(min)),
+        maxoptions: (value, max) => value === '' || (Array.isArray(value) && value.length <= parseFloat(max)),
+    },
+    is(value, rules = []) {
+        for (const index in rules) {
+            const rule = rules[index].split(':')
+            const test = this.tests[rule[0]] || window.AlpineValidationRules[rule[0]]
+
+            if (!test) {
+                throw Error('Invalid rule in validator: ' + rule[0])
+            }
+
+            const result = test.apply(this, [value, rule[1]])
+
+            if (!result) return rules[index]
+        }
+
+        return true
+    },
+}
+
+const AlpineValidateCustomDirective = {
+    start() {
+        checkForAlpine()
+
+        Alpine.addMagicProperty('invalid', ($el) => {
+            return (target, errorType = '') => {
+                const originalTarget = target
+
+                // Support for CSS query selector
+                if (typeof target === 'string') {
+                    target = document.querySelector(target)
+                }
+
+                if (!(target instanceof Element)) {
+                    throw Error('Unsupported $invalid target: ', originalTarget)
+                }
+
+                if (target.$dirty !== true) {
+                    return false
+                }
+
+                // Note Alpine has a bug end it strips all backslashes breaking regular expresions
+
+                return (errorType === '')
+                    ? target.$valid !== true
+                    : (
+                        // Note Alpine has a bug end it strips all backslashes breaking regular expresions
+                        (typeof target.$valid === 'string')
+                        ? target.$valid.replace(/\\/g, '') === errorType
+                        : target.$valid === errorType
+                    )
+            }
+        })
+
+        Alpine.onBeforeComponentInitialized((component) => {
```

Comment:
Could we maybe wrap all of this up into a small helper function? The part that handles whether or not the element has the specified attribute, etc.

Reply:
Sure, we can send another PR to master and then update this one to keep the feature separated. I won't have time to work on it this week though but if any of you guys have time to create the helper, feel free to give it a go.